streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java

```diff
@@ -435,7 +435,7 @@ public void shouldNotPunctuateSystemTimeWhenIntervalNotElapsed() {
         task = createStatelessTask(false);
         task.initialize();
         long now = time.milliseconds();
-        assertTrue(task.maybePunctuateSystemTime()); // first time we always punctuate
```

Comment:
@mihbor this makes little sense to me, is there anything I am missing? If users ask for punctuation in a certain amount of time, why punctuate immediately?

Reply:
TBH I don't remember right now any reason for this. Probably just kept it consistent with stream time punctuation. Right now, I can't think of any reason why punctuating immediately would be desirable though.

Reply:
The consistency point is true IIRC -- for `streamTimePunctuationQueue.schedule(schedule);` we also punctuate immediately -- there is hard to change as we would need to consider the timestamp of the very first record we process.

I am just thinking, if we should keep the current behavior and put the burden on the user to ignore the very first punctuation call? We had a similar ticket in the passed already, but no conclusion how to proceed: https://issues.apache.org/jira/browse/KAFKA-6092

Reply:
In `maybePunctuateStreamTime()` I see no immediate punctuation unless there is a valid timestamp for each partition. That said, the situation is different once each partition has a valid timestamp and I see your point.

Some context... I hit that problem because I am using `context.schedule(..., WALL_CLOCK_TIME, ...)` more like a `scheduleOnce()` i.e., when the punctuation is triggered, I immediately cancel the scheduler and schedule a new one because I am dealing with variable intervals. In my use case an immediate punctuation makes no sense, however I agree that this is a question of expectations. Some documentation changes could explain that aspect more clearly and help not creating wrong expectations, I'll push another PR for that.

I currently worked around this issue by ignoring all punctuation calls before the expected system time. The result is that in my case I ignore 1 out of every 2 punctuations. This is no big deal as my intervals are long (minutes to hours).

Reply:
> I ignore 1 out of every 2 punctuations

Uh. That's kinda painful... I think we need to discuss this in more details, ie, what semantics we want to provide. \cc @bbejeck @dguy @guozhangwang @miguno

Reply:
Hmm, interesting. In my head it makes sense for this case to not punctuate until after the time has elapsed rather than immediately. But i agree that it then wouldn't be consistent with stream time punctuation, which would be confusing. I'm afraid i can't think of a good answer!

Reply:
@fredfp @dguy I left some comments on the ticket itself: https://issues.apache.org/jira/browse/KAFKA-6323. My take is that we should fix the code logic to be more reasonable, than explaining the current behavior as a doc change. In a word I think it is OK to make each type's punctuation semantics a bit different but more straight-forward for users to understand than trying to make sure they are consistent, because one type is really "data driven" while the other is time driven, so their semantics could be different by design.

LMK what do you think of the behavior change proposal.

Reply:
taken care of in https://github.com/apache/kafka/pull/4301/commits/5ae6dd9815f86158fc98202a8216d22af6d78d47